File: course_3/module_4/programming_assignment_4/solution_2.py
class Item:
    """Represents an item with a value and weight."""
    def __init__(self, value: int, weight: int):
        self.value = value
        self.weight = weight


class Knapsack:
    """Represents a knapsack with a given capacity and a list of items."""
    def __init__(self, capacity: int, items: list[Item]):
        self.capacity = capacity
        self.items = items

    def solve(self) -> int:
        """Solves the knapsack problem using optimizations."""
        dp = [0] * (self.capacity + 1)

        # Sort items by value-to-weight ratio.
        self.items.sort(key=lambda x: x.value / x.weight, reverse=True)

        # Update dp array for each item.
        for item in self.items:
            for w in range(self.capacity, item.weight - 1, -1):
                dp[w] = max(dp[w], dp[w - item.weight] + item.value)

        return dp[self.capacity]


def filter_items(items: list[Item], capacity: int) -> list[Item]:
    """Filters out irrelevant items."""
    return [item for item in items if item.weight <= capacity]


def read_input_file(file_path: str) -> Knapsack:
    """Reads the input file and constructs a Knapsack object."""
    with open(file_path, 'r') as file:
        lines = file.readlines()

    first_line = lines[0].split()
    capacity = int(first_line[0])

    items = []
    for line in lines[1:]:
        value, weight = map(int, line.split())
        items.append(Item(value, weight))

    # Filter items and return Knapsack instance.
    items = filter_items(items, capacity)
    return Knapsack(capacity, items)


if __name__ == "__main__":
    file_path = "knapsack_big.txt"
    knapsack = read_input_file(file_path)
    optimal_value = knapsack.solve()
    print(f"The optimal value is: {optimal_value}.")

File: course_3/module_4/programming_assignment_4/solution_2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Item represents an item with a value and weight.
type Item struct {
	Value  int
	Weight int
}

// Knapsack represents a knapsack with a given capacity and a list of items.
type Knapsack struct {
	Capacity int
	Items    []Item
}

// NewKnapsack creates a new knapsack with a given capacity.
func NewKnapsack(capacity int) *Knapsack {
	return &Knapsack{
		Capacity: capacity,
		Items:    []Item{},
	}
}

// AddItem adds an item to the knapsack.
func (k *Knapsack) AddItem(item Item) {
	k.Items = append(k.Items, item)
}

// Solve solves the knapsack problem using dynamic programming.
func (k *Knapsack) Solve() int {
	numItems := len(k.Items)
	// Create a 2D DP table initialized with zeros.
	dp := make([][]int, numItems+1)
	for i := range dp {
		dp[i] = make([]int, k.Capacity+1)
	}

	// Fill the DP table.
	for i := 1; i <= numItems; i++ {
		for w := 0; w <= k.Capacity; w++ {
			if k.Items[i-1].Weight > w {
				dp[i][w] = dp[i-1][w] // Item cannot be included.
			} else {
				// Take the maximum of including or excluding the item.
				dp[i][w] = max(dp[i-1][w], dp[i-1][w-k.Items[i-1].Weight]+k.Items[i-1].Value)
			}
		}
	}

	return dp[numItems][k.Capacity]
}

// max returns the maximum of two integers.
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// ReadInputFile reads the input file and constructs a Knapsack object.
func ReadInputFile(filePath string) (*Knapsack, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	// Read the first line for capacity and number of items.
	scanner.Scan()
	firstLine := strings.Fields(scanner.Text())
	capacity, _ := strconv.Atoi(firstLine[0])

	// Create a new knapsack.
	knapsack := NewKnapsack(capacity)

	// Read the items from the remaining lines.
	for scanner.Scan() {
		line := strings.Fields(scanner.Text())
		value, _ := strconv.Atoi(line[0])
		weight, _ := strconv.Atoi(line[1])
		knapsack.AddItem(Item{Value: value, Weight: weight})
	}

	return knapsack, nil
}

func main() {
	// Path to the input file.
	filePath := "course_3/module_4/programming_assignment_4/knapsack_big.txt"

	// Read the knapsack data from the file.
	knapsack, err := ReadInputFile(filePath)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	// Solve the knapsack problem.
	optimalValue := knapsack.Solve()
	fmt.Printf("The optimal value is: %d\n", optimalValue)
}
